### PR TITLE
Refactor with new framework

### PR DESCRIPTION
**Pull request by zzclynn**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -58,32 +58,28 @@ The ModelScope Notebook offers a free-tier that allows ModelScope user to run th
 
 ## Quickstart
 
-To use modelscope-agent, all you need is to instantiate an `AgentExecutor` object, and use `run()` to execute your task. For faster agent implementation, please refer to [demo_agent](demo/demo_qwen_agent.ipynb). Online demo is available on [ModelScope](https://modelscope.cn/studios/damo/ModelScopeGPT/summary)
+To use modelscope-agent, all you need is to instantiate an `Agent` object, and use `run()` to execute your task. For faster agent implementation, please refer to [demo_agent](demo/demo_qwen_agent.ipynb). Online demo is available on [ModelScope](https://modelscope.cn/studios/damo/ModelScopeGPT/summary)
```
Comment (wenmengzhou) [inline README.md:61]: change the demo to agentfabric/agentscope

File: modelscope_agent/agent.py [hunk 2]
```diff
@@ -1,377 +1,174 @@
...
-            action_parser (Optional[ActionParser], optional): this module is responsible for parsing output of llm
-            to executable actions. Defaults to use MsActionParser.
-            tool_retrieval (Optional[Union[bool, ToolRetrieval]], optional): Retrieve related tools by input task,
-            since most of the tools may be useless for LLM in specific task.
-            If it is bool type and is True, will use default tool_retrieval. Defaults to True.
-            knowledge_retrieval (Optional[KnowledgeRetrieval], optional): If user want to use extra knowledge,
-            this component can be used to retrieve related knowledge. Defaults to None.
+        init tools/llm for one agent
+
+        :param function_list: Optional[List[Union[str, Dict]]] :
```
Comment (wenmengzhou) [inline modelscope_agent/agent.py]: doc str format should be the same as before

File: modelscope_agent/agents/agent_builder.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+ANSWER = 'Answer: '
+CONFIG = 'Config: '
+RICH_CONFIG = 'RichConfig: '
+ASSISTANT_TEMPLATE = """Answer: {answer}
+Config: {config}
+RichConfig: {rich_config}"""
+
+
+class AgentBuilder(Agent):
+
```
Comment (wenmengzhou) [inline modelscope_agent/agents/agent_builder.py:82]: add  docstr for this agent

File: modelscope_agent/llm/base.py [hunk 4]
```diff
@@ -1,64 +1,125 @@
...
+            else:
+                return True
+        except Exception:
+            return False
+
+    def chat_with_raw_prompt(self,
+                             prompt: str,
+                             stop: Optional[List[str]] = None,
+                             **kwargs) -> str:
+        raise '[Do not Support]'
```
Comment (wenmengzhou) [inline modelscope_agent/llm/base.py]: raise NotImplementedError and. catch NotImplementedError. in support_raw_prompt

File: modelscope_agent/llm/base.py [hunk 5]
```diff
@@ -1,64 +1,125 @@
...
-        Args:
-            input_text (str): input text
-        Returns:
-            list[int]: token_ids
-        """
-        raise NotImplementedError
 
-    def detokenize(self, input_ids: List[int]) -> str:
-        """detokenize
+class BaseChatModel(ABC):
```
Comment (wenmengzhou) [inline modelscope_agent/llm/base.py:27]: this class is so important that each function should be added with detailed docstr to illustrate the functionality

File: modelscope_agent/agent.py [hunk 6]
```diff
@@ -1,377 +1,182 @@
...
+        self.function_list = []
+        self.function_map = {}
+        if function_list:
+            for function in function_list:
+                self._register_tool(function)
+                try:
+                    self._register_tool(function)
+                except Exception as e:
+                    logger.query_warning(
+                        uuid=kwargs.get('uuid_str', 'local_user'),
```
Comment (wenmengzhou) [inline modelscope_agent/agent.py]: why we need to print uuid_str inside a framework?

Comment (wenmengzhou) [reply modelscope_agent/agent.py]: we should raise the exception and let the upper apps to handle and print the right info

File: modelscope_agent/agent.py [hunk 7]
```diff
@@ -1,377 +1,182 @@
...
-        clear history and agent state
-        """
-        self.prompt_generator.reset()
-        self.agent_state = {}
+        A built-in tool call detection: After encapsulating function calls in the LLM layer, this is no longer needed
 
-    def parse_action_args(self, action_args):
         """
-        replace action_args in str to Image/Video/Audio Wrapper, so that tool can handle them
+        special_func_token = '\nAction:'
```
Comment (wenmengzhou) [inline modelscope_agent/agent.py]: this logic is related to what the prompt is  and looks a little bit tricky

File: modelscope_agent/agent.py [hunk 8]
```diff
@@ -1,377 +1,182 @@
...
-                    self.prompt_generator.reset()
-                    return
-            else:
-                exec_result = f"Unknown action: '{action}'. "
-                yield {'exec_result': exec_result}
-                self.prompt_generator.reset()
-                return
-
-    def reset(self):
+    def _detect_tool_by_special_token(self, text: str):
```
Comment (wenmengzhou) [inline modelscope_agent/agent.py]: a better name may be  `extract_tool_info`

File: modelscope_agent/agent.py [hunk 9]
```diff
@@ -1,377 +1,182 @@
...
+    def receive(self,
+                message: Union[Dict, str],
+                sender: 'Agent',
+                request_reply: Optional[bool] = None,
+                **kwargs):
+        if request_reply is False or request_reply is None:
+            return
+        reply = self.run(message, sender=sender, **kwargs)
+        if reply is not None:
+            self.send(reply, sender, **kwargs)
```
Comment (wenmengzhou) [inline modelscope_agent/agent.py]: these two interface are not used and referenced, may be added later, delete them now

File: modelscope_agent/agents/gen_keyword.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+Keywords:{{"keywords_zh": ["核心公式", "公式"], "keywords_en": ["core formula", "formula", "equation"]}}
+
+Question:{user_request}
+Keywords:
+"""
+
+PROMPT_TEMPLATE_EN = """Please extract keywords from the question, both in Chinese and English,
+and supplement them appropriately with relevant keywords that are not in the question.
+Try to divide keywords into verb/noun/adjective types and avoid long phrases. Keywords are provided in JSON format,
+such as {{"keywords_zh": ["关键词1", "关键词2"], "keywords_en": ["keyword 1", "keyword 2"]}}
```
Comment (wenmengzhou) [inline modelscope_agent/agents/gen_keyword.py:24]: why are there chinese words in english prompt template?

File: modelscope_agent/agents/gen_keyword.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+Keywords:
+"""
+
+PROMPT_TEMPLATE = {
+    'zh': PROMPT_TEMPLATE_ZH,
+    'en': PROMPT_TEMPLATE_EN,
+}
+
+
+class GenKeyword(Agent):
```
Comment (wenmengzhou) [inline modelscope_agent/agents/gen_keyword.py:45]: KeywordExtractor is a better name for agent, and it is better to give the usage and input output example in docstr

File: modelscope_agent/agents/react.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
+from typing import Dict, Iterator, List, Optional
```
Comment (wenmengzhou) [inline modelscope_agent/agents/react.py]: add doc to clarify the difference between react and react_chat,  which probably one use chat format messages, and one use raw prompt?

File: modelscope_agent/agents/role_play.py [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+        return (func_name is not None), func_name, func_args, text
+
+    def _parse_role_config(self, config: dict, lang: str = 'zh') -> str:
+        if lang == 'en':
+            return self._parse_role_config_en(config)
+        else:
+            return self._parse_role_config_zh(config)
+
+    def _parse_role_config_en(self, config: dict) -> str:
+        prompt = 'You are playing as an AI-Agent, '
```
Comment (wenmengzhou) [inline modelscope_agent/agents/role_play.py:300]: add docstr for config

File: modelscope_agent/llm/custom.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+import json
+import requests
+
+from .base import BaseChatModel, register_llm
+
+
+@register_llm('custom')
+class CustomLLM(BaseChatModel):
+    '''
+        the llm service from http.
```
Comment (wenmengzhou) [inline modelscope_agent/llm/custom.py]: make it clear what is the api schema of the backeend service

File: modelscope_agent/llm/custom.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        super().__init__(model, model_server)
+
+        self.token = os.getenv('HTTP_LLM_TOKEN', None)
+        self.model = os.getenv('HTTP_LLM_MODEL', None)
+        self.url = os.getenv('HTTP_LLM_URL', None)
+
+        if self.token is None:
+            raise ValueError('HTTP_LLM_TOKEN is not set')
+
+    def http_request(self, data):
```
Comment (wenmengzhou) [inline modelscope_agent/llm/custom.py]: http_request -> _http_request

File: modelscope_agent/llm/custom.py (hunk 14 above)
Comment (wenmengzhou) [reply modelscope_agent/llm/custom.py]: input and output

File: modelscope_agent/llm/custom.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        data = {'model': self.model, 'messages': messages, 'n': 1}
+        message = self._inference(data)
+        return message['content']
+
+    def _chat_stream(self,
+                     messages: List[Dict],
+                     stop: Optional[List[str]] = None,
+                     **kwargs) -> Iterator[str]:
+        # todo: implement the streaming chat
+        raise NotImplementedError
```
Comment (wenmengzhou) [inline modelscope_agent/llm/custom.py]: stream example should also be added later, a todo list is needed

File: modelscope_agent/llm/custom.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        if 'function_call' in message and message['function_call'] != {}:
+            new_message['function_call'] = message.get('function_call')
+
+        return new_message
+
+    def _inference(self, data: dict) -> dict:
+        retry_count = 0
+        max_retries = 3
+        message = {'content': ''}
+        while retry_count <= max_retries:
```
Comment (wenmengzhou) [inline modelscope_agent/llm/custom.py]: a general retry dectorator should be added to modelscope agent for retry

File: modelscope_agent/llm/dashscope.py [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+    last_len = 0
+    delay_len = 5
+    in_delay = False
+    text = ''
+    for trunk in response:
+        if trunk.status_code == HTTPStatus.OK:
+            # logger at the first for the request_id, and the last time for whole output
+            if not text or trunk.output.choices[0].finish_reason != 'null':
+                logger.query_info(
+                    uuid=kwargs.get('uuid_str', ''),
```
Comment (wenmengzhou) [inline modelscope_agent/llm/dashscope.py]: again. why should we need uuid_str?

File: modelscope_agent/llm/dashscope.py [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+                in_delay = True
+                continue
+            else:
+                in_delay = False
+                real_text = text[:-delay_len]
+                now_rsp = real_text[last_len:]
+                yield now_rsp
+                last_len = len(real_text)
+        else:
+            logger.query_error(
```
Comment (wenmengzhou) [inline modelscope_agent/llm/dashscope.py:38]: raise exception for better error handling

File: modelscope_agent/llm/modelscope.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+        model = Swift.from_pretrained(model, self.lora_ckpt_dir)
+        self.model = model
+
+    def _chat_stream(self,
+                     messages: List[Dict],
+                     stop: Optional[List[str]] = None,
+                     **kwargs) -> Iterator[str]:
+        # agents = messages[0]['content']
+        # todo: implement the streaming chat
+        raise NotImplementedError
```
Comment (wenmengzhou) [inline modelscope_agent/llm/modelscope.py:91]: chat stream should be implemented later, add a todo item

Comment (wenmengzhou) [reply modelscope_agent/llm/modelscope.py:91]: one probable tool is https://github.com/LowinLi/transformers-stream-generator

File: modelscope_agent/memory/base.py [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+            history: List of Message
+
+        Returns: None
+
+        """
+        directory = os.path.dirname(self.path)
+        if not os.path.exists(directory):
+            os.makedirs(directory)
+
+        with open(self.path, 'w', encoding='utf-8') as file:
```
Comment (wenmengzhou) [inline modelscope_agent/memory/base.py:29]: looks weird,  save_memory does not save self.history?   save_memory  load memory. get_history. update_history. seems unrelated between memory and history

File: modelscope_agent/memory/memory_with_file_knowledge.py [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+        try:
+            keyword_dict = json5.loads(keyword)
+            keyword_dict['text'] = query
+            query_with_keyword = keyword_dict
+        except Exception:
+            query_with_keyword = query
+
+        # retrieval related content
+        records = [
+            RefMaterialInput(**record) for record in json5.loads(records)
```
Comment (wenmengzhou) [inline modelscope_agent/memory/memory_with_file_knowledge.py]: seems called json5.loads repeatedly, line 53 has already done this

File: modelscope_agent/storage/base.py [hunk 23]
```diff
@@ -0,0 +1,14 @@
+from abc import ABC, abstractmethod
+
+
+class BaseStorage(ABC):
+
+    @abstractmethod
```
Comment (wenmengzhou) [inline modelscope_agent/storage/base.py:6]: where is the init interface? some storage need to be initialized with some resources

File: modelscope_agent/storage/base.py [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+    def add(self, *args, **kwargs):
+        """add items to db or indexer"""
+        pass
+
+    @abstractmethod
+    def search(self, *args, **kwargs):
+        """search from db or indexer"""
```
Comment (wenmengzhou) [inline modelscope_agent/storage/base.py:13]: delete interface may be needed

File: modelscope_agent/schemas.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+    time: str
+    source: str
+    raw: list
+    title: str
+    topic: str
+    checked: bool
+    session: list
+
+
+class AgentHolder(BaseModel):
```
Comment (wenmengzhou) [inline modelscope_agent/schemas.py]: seems strange,  AgentAttr?

File: modelscope_agent/tools/modelscope_tools/pipeline_tool.py [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+        # only initialize when this tool is really called to save memory
+        if not self.is_initialized:
+            self.pipeline = pipeline(
+                task=self.task,
+                model=self.model,
+                model_revision=self.model_revision,
+                **self.pipeline_params)
+        self.is_initialized = True
+
+    def call(self, params: str, **kwargs) -> str:
```
Comment (wenmengzhou) [inline modelscope_agent/tools/modelscope_tools/pipeline_tool.py:57]: call should receive a arg called `remote` to do switch between local and remote call?

File: modelscope_agent/tools/utils/output_wrapper.py [hunk 27]
```diff
@@ -137,20 +140,19 @@ def __init__(self, video) -> None:
...
             raise TypeError(
                 'Current only support load from filepath when it is video')
 
         self._raw_data = video
-        self._repr = f'![IMAGEGEN]({self._path})'
+
+        if self._path.endswith('.gif'):
+            self._repr = f'![IMAGEGEN]({self._path})'
+        else:
+            self._repr = f'<audio src="{self._path}"/>'
```
Comment (wenmengzhou) [inline modelscope_agent/tools/utils/output_wrapper.py]: audio -> video?

**Review by wenmengzhou — CHANGES_REQUESTED** [review]
changes should be made according to code review above

File: modelscope_agent/agents/react.py (hunk 12 above)
Comment (tuhahaha) [reply modelscope_agent/agents/react.py]: Yes, maybe I should delete them, they haven't been used yet

File: modelscope_agent/tools/modelscope_tools/pipeline_tool.py (hunk 26 above)
Comment (tuhahaha) [reply modelscope_agent/tools/modelscope_tools/pipeline_tool.py:57]: It's strange to let the agent choose remote or local. Perhaps the default solution can be implemented in the tool, or try local, except remote?

File: modelscope_agent/agent.py (hunk 7 above)
Comment (tuhahaha) [reply modelscope_agent/agent.py]: Deleted

File: modelscope_agent/agent.py (hunk 8 above)
Comment (tuhahaha) [reply modelscope_agent/agent.py]: Deleted

File: modelscope_agent/agent.py (hunk 9 above)
Comment (tuhahaha) [reply modelscope_agent/agent.py]: Deleted

File: modelscope_agent/agents/role_play.py (hunk 13 above)
Comment (tuhahaha) [reply modelscope_agent/agents/role_play.py:300]: The latest version no longer includes this config, but instead passes in a str instruction after external processing. At present, I will annotate it with docstr first

File: modelscope_agent/agent.py (hunk 6 above)
Comment (zzhangpurdue) [reply modelscope_agent/agent.py]: removed

File: modelscope_agent/llm/custom.py (hunk 14 above)
Comment (zzhangpurdue) [reply modelscope_agent/llm/custom.py]: removed the entire customLLM, should be replaced by fastchat or llamaindex later

File: modelscope_agent/llm/custom.py (hunk 17 above)
Comment (zzhangpurdue) [reply modelscope_agent/llm/custom.py]: added

File: modelscope_agent/llm/dashscope.py (hunk 18 above)
Comment (zzhangpurdue) [reply modelscope_agent/llm/dashscope.py]: removed

File: modelscope_agent/agents/role_play.py (hunk 13 above)
Comment (wenmengzhou) [reply modelscope_agent/agents/role_play.py:300]: @tuhahaha  in which place can I see the latest version?  Maybe I did'nt catch all the details you said above

File: modelscope_agent/llm/modelscope.py (hunk 20 above)
Comment (wenmengzhou) [reply modelscope_agent/llm/modelscope.py:91]: todo item format is  
```
# TODO  @zzclynn add stream chat

```

File: modelscope_agent/tools/modelscope_tools/pipeline_tool.py (hunk 26 above)
Comment (wenmengzhou) [reply modelscope_agent/tools/modelscope_tools/pipeline_tool.py:57]: yeah. I agree, maybe modelscope_hub_tool should implements the local call and modelscope_api_tool should implements the remote call

File: modelscope_agent/agents/role_play.py (hunk 13 above)
Comment (tuhahaha) [reply modelscope_agent/agents/role_play.py:300]: [这里](https://github.com/QwenLM/Qwen-Agent)哈，经常会涉及大的调整，还不是最稳定的版本，我在持续更新中。

File: modelscope_agent/tools/utils/output_wrapper.py (hunk 27 above)
Comment (suluyana) [reply modelscope_agent/tools/utils/output_wrapper.py]: done

File: modelscope_agent/schemas.py (hunk 25 above)
Comment (zzhangpurdue) [reply modelscope_agent/schemas.py]: done

File: modelscope_agent/storage/base.py (hunk 24 above)
Comment (zzhangpurdue) [reply modelscope_agent/storage/base.py:13]: added

File: modelscope_agent/storage/base.py (hunk 23 above)
Comment (zzhangpurdue) [reply modelscope_agent/storage/base.py:6]: init method might be varied from one db to other dbs. 
no common attribute at this moment.

File: modelscope_agent/memory/memory_with_file_knowledge.py (hunk 22 above)
Comment (zzhangpurdue) [reply modelscope_agent/memory/memory_with_file_knowledge.py]: removed

File: modelscope_agent/memory/base.py (hunk 21 above)
Comment (zzhangpurdue) [reply modelscope_agent/memory/base.py:29]: has changed to reasonable name

**Review by wenmengzhou — APPROVED** [review]

**Pull request merged**